crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -298,8 +359,8 @@ impl<'db> InnerIntersectionBuilder<'db> {
                     return;
                 }
             }
-            for index in to_remove.iter().rev() {
-                self.positive.swap_remove_index(*index);
+            for index in to_remove {
+                self.positive.swap_remove_index(index);
```

Comment:
I think it's now possible that `index` is incorrect because `FxHashSet` doesn't preserve insertion order. That means, it can now happen that the index order is `3, 1, 5` 

I'm also a bit confused why it's no longer necessary to iterate in reverse order?

Reply:
Oh, really good point. Yeah, this is incorrect. I think we have some missing test coverage here.

Reply:
I wonder if it's worth using a `BTreeSet` or if we could keep using a `Vec` and filter out duplicate values when iterating (by simply storing the last removed index and skipping if it remains the same). 

I don't know which one will perform better. That probably depends on how common duplicates are

Reply:
`to_remove` might have been populated by multiple different loops over `self.positive` and `self.negative` with my refactor, so this isn't just about deduplicating. It's also about ensuring the collection stays in sorted order.

I could instead use a single `smallvec` for each loop, which would look a bit closer to the current code. It would be a bit more repetitive but might be more performant due to fewer allocations 🤷

Reply:
On closer examination, I think the `BTreeSet` was both unnecessary and possibly buggy... I've removed it. Thanks for pushing me on this :-)